### PR TITLE
[Auto Techsupport] Event driven Techsupport Changes

### PR DESCRIPTION
**Pull request by vivekreddynv**
#### What I did

sonic-utilities changes required for feature "Event Driven TechSupport Invocation & CoreDump Mgmt". [HLD](https://github.com/Azure/SONiC/pull/818 )

Summary of the changes:

- Added the AUTO GEN CLI for the CFG DB tables required for this feature
- Added the coredump_gen_handler.py & techsupport_cleanup.py scripts.
- Added the UT's required for these scripts.
- Enhanced coredump-compress & generate-dump scripts

#### How to verify it

```
tests/coredump_gen_handler_test.py::TestCoreDumpCreationEvent::test_core_dump_cleanup PASSED [ 22%]
tests/coredump_gen_handler_test.py::TestCoreDumpCreationEvent::test_core_dump_with_exit_event_unknown_cmd PASSED [ 22%]
tests/coredump_gen_handler_test.py::TestCoreDumpCreationEvent::test_core_dump_with_no_exit_event PASSED [ 22%]
tests/coredump_gen_handler_test.py::TestCoreDumpCreationEvent::test_feature_table_not_set PASSED [ 22%]
tests/coredump_gen_handler_test.py::TestCoreDumpCreationEvent::test_global_rate_limit_interval PASSED [ 22%]
tests/coredump_gen_handler_test.py::TestCoreDumpCreationEvent::test_invalid_since_argument PASSED [ 22%]
tests/coredump_gen_handler_test.py::TestCoreDumpCreationEvent::test_invoc_ts_after_rate_limit_interval PASSED [ 22%]
tests/coredump_gen_handler_test.py::TestCoreDumpCreationEvent::test_invoc_ts_state_db_update PASSED [ 22%]
tests/coredump_gen_handler_test.py::TestCoreDumpCreationEvent::test_max_core_size_limit_not_crossed PASSED [ 22%]
tests/coredump_gen_handler_test.py::TestCoreDumpCreationEvent::test_per_proc_rate_limit_interval PASSED [ 22%]
tests/coredump_gen_handler_test.py::TestCoreDumpCreationEvent::test_since_argument PASSED [ 22%]
tests/techsupport_cleanup_test.py::TestTechsupportCreationEvent::test_dump_cleanup PASSED [ 79%]
tests/techsupport_cleanup_test.py::TestTechsupportCreationEvent::test_no_cleanup_state_disabled PASSED [ 79%]
tests/techsupport_cleanup_test.py::TestTechsupportCreationEvent::test_no_cleanup_state_enabled PASSED [ 80%]
tests/techsupport_cleanup_test.py::TestTechsupportCreationEvent::test_state_db_update PASSED [ 80%]
```

#### New command output (if the output of a command-line utility has changed)

```

admin@sonic:~$ show auto-techsupport global
STATE      RATE LIMIT INTERVAL    MAX TECHSUPPORT SIZE    MAX CORE SIZE  SINCE
-------  ---------------------  ----------------------  ---------------  ----------
enabled                    180                      10                5  2 days ago

admin@sonic:~$ show auto-techsupport-feature 
FEATURE NAME    STATE       RATE LIMIT INTERVAL
--------------  --------  ---------------------
bgp             enabled                     600
database        enabled                     600
dhcp_relay      enabled                     600
lldp            enabled                     600
macsec          enabled                     600
mgmt-framework  enabled                     600
nat             enabled                     600
pmon            enabled                     600
radv            enabled                     600
restapi         disabled                    800
sflow           enabled                     600
snmp            enabled                     600
swss            disabled                    800
syncd           enabled                     600
teamd           enabled                     600
telemetry       enabled                     600
admin@sonic:~$ show auto-techsupport history

TECHSUPPORT DUMP                          TRIGGERED BY    CORE DUMP
----------------------------------------  --------------  -----------------------------
sonic_dump_r-lionfish-16_20210901_221402  bgpcfgd         bgpcfgd.1630534439.55.core.gz
sonic_dump_r-lionfish-16_20210901_203725  snmp-subagent   python3.1630528642.23.core.gz
sonic_dump_r-lionfish-16_20210901_222408  lldpmgrd        python3.1630535045.34.core.gz

### CODE REVIEW

File: scripts/coredump_gen_handler.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+            ts_dump = ts_key.split("|")[-1]
+            if container_name and container_name not in self.core_ts_map:
+                self.core_ts_map[container_name] = []
+            self.core_ts_map[container_name].append((int(creation_time), ts_dump))
+        for container_name in self.core_ts_map:
+            self.core_ts_map[container_name].sort()
+
+    def fetch_exit_event(self):
+        """Fetch the relevant entry in the AUTO_TECHSUPPORT|PROC_EXIT_EVENTS table"""
+        comm, _, pid, _, _ = self.core_name.split(".")
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply scripts/coredump_gen_handler.py:165]: Sorry, this method is redundant. Will remove this

File: scripts/coredump_gen_handler.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+            return SINCE_DEFAULT
+        rc, _, stderr = subprocess_exec(["date", "--date='{}'".format(since_cfg)])
+        if rc == 0:
+            return since_cfg
+        return SINCE_DEFAULT
+
+    def invoke_ts_cmd(self, since_cfg):
+        since_cfg = "'" + since_cfg + "'"
+        cmd  = " ".join(["show", "techsupport", "--since", since_cfg])
+        _, _, _ = subprocess_exec(["show", "techsupport", "--since", since_cfg])
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply scripts/coredump_gen_handler.py:112]: Exit code returned by this execution might not reflect if the tech-support dump creation went properly or not. Thus checking if the dump was indeed created.

[the remark this answers is not in the document]
Comment (ganglyu) [reply scripts/coredump_gen_handler.py:112]: Do you mean tech-support dump can be created even exit code is 1?

**Comment by ganglyu** [comment]
@vivekreddynv, you can check the result of subprocess_exec, or just remove "_, _, _"

**Comment by vivekreddynv** [comment]
> @vivekreddynv, you can check the result of subprocess_exec, or just remove "_, _, _"

Sure, will remove the "_, _, _"

**Review by qiluo-msft — CHANGES_REQUESTED** [review]
As commments

File: scripts/coredump-compress [hunk 3]
```diff
@@ -7,11 +7,26 @@ while [[ $# > 1 ]]; do
...
     fi
 fi
 
 /bin/gzip -1 - > /var/core/${PREFIX}core.gz
+
+if [[ ! -z $CONTAINER_ID ]]; then
+   CONTAINER_NAME=$(docker inspect --format='{{.Name}}' ${CONTAINER_ID} | cut -c2-)
+   if [[ ! -z ${CONTAINER_NAME} ]]; then
+        # coredump_gen_handler invokes techsupport if all the other required conditions  are met
+        # explicitly passing in the env vars because coredump-compress's namespace doesn't have these set by default
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply scripts/coredump-compress:25]: Do you mean to suggest that these should be passed from the parent process that is calling coredump-compress? Is yes, then that is not possible as this is invoked by kernel directly

File: scripts/coredump-compress [hunk 4]
```diff
@@ -7,11 +7,26 @@ while [[ $# > 1 ]]; do
...
+        setsid $(echo > /tmp/coredump_gen_handler.log;
+             export PYTHONPATH=$PYTHONPATH:/usr/local/lib/python3.7/dist-packages/;
+             export PATH=$PATH:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin;
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply scripts/coredump-compress:28]: PATH and PYTHONPATH are required for the genearte_dump script (which is called inside the coredump_gen_handler.py script) to find the required binaries and scripts

File: scripts/coredump-compress [hunk 5]
```diff
@@ -7,11 +7,26 @@ while [[ $# > 1 ]]; do
...
 
 /bin/gzip -1 - > /var/core/${PREFIX}core.gz
+
+if [[ ! -z $CONTAINER_ID ]]; then
+   CONTAINER_NAME=$(docker inspect --format='{{.Name}}' ${CONTAINER_ID} | cut -c2-)
+   if [[ ! -z ${CONTAINER_NAME} ]]; then
+        # coredump_gen_handler invokes techsupport if all the other required conditions  are met
+        # explicitly passing in the env vars because coredump-compress's namespace doesn't have these set by default
+        setsid $(echo > /tmp/coredump_gen_handler.log;
+             export PYTHONPATH=$PYTHONPATH:/usr/local/lib/python3.7/dist-packages/;
```
[the remark this answers is not in the document]
Comment (vivekreddynv) [reply scripts/coredump-compress:27]: Should i also add python ver 2? I'm not sure of any python2 scripts are used in the generate_dump script but I'll add that , gud to have it nevertheless

**Comment by liat-grozovik** [comment]
@qiluo-msft and @ganglyu could you please review latest code changes?

File: scripts/coredump-compress [hunk 6]
```diff
@@ -7,11 +7,28 @@ while [[ $# > 1 ]]; do
...
 fi
 
 /bin/gzip -1 - > /var/core/${PREFIX}core.gz
+
+if [[ ! -z $CONTAINER_ID ]]; then
+   CONTAINER_NAME=$(docker inspect --format='{{.Name}}' ${CONTAINER_ID} | cut -c2-)
+   if [[ ! -z ${CONTAINER_NAME} ]]; then
+      # coredump_gen_handler invokes techsupport if all the other required conditions are met
+      # explicitly passing in the env vars because coredump-compress's namespace doesn't have these set by default
+      for path in $(find /usr/local/lib/python[^2]*/dist-packages -maxdepth 0); do
```
Comment (qiluo-msft) [inline scripts/coredump-compress:26]: >^2 [](http://example.com/codeflow?start=47&length=2)

check `python` and `python3*` only?

File: scripts/coredump-compress [hunk 7]
```diff
@@ -7,11 +7,28 @@ while [[ $# > 1 ]]; do
...
+          PYTHONPATH=$PYTHONPATH:$path
+      done
+      setsid $(echo > /tmp/coredump_gen_handler.log;
+             export PYTHONPATH=$PYTHONPATH;
+             python3 /usr/local/bin/coredump_gen_handler.py ${PREFIX}core.gz ${CONTAINER_NAME} &>> /tmp/coredump_gen_handler.log) &
```
Comment (qiluo-msft) [inline scripts/coredump-compress:31]: >python3 [](http://example.com/codeflow?start=13&length=7)

If add a shebang to the py file, do you still need to use `python3` and `PYTHONPATH`?

File: scripts/coredump-compress (hunk 6 above)
Comment (vivekreddynv) [reply scripts/coredump-compress:26]: Only python3.x, Can't involve both python2 & python3 both. imports won't work if i do so

File: scripts/coredump-compress (hunk 7 above)
Comment (vivekreddynv) [reply scripts/coredump-compress:31]: i can avoid python3 but PYTHONPATH is still required.

File: show/plugins/auto_techsupport.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+             cls=clicommon.AliasedGroup)
+def AUTO_TECHSUPPORT():
+    """ AUTO_TECHSUPPORT part of config_db.json """
+
+    pass
+
+
+@AUTO_TECHSUPPORT.command(name="global")
+@clicommon.pass_db
+def AUTO_TECHSUPPORT_GLOBAL(db):
```
Comment (qiluo-msft) [inline show/plugins/auto_techsupport.py:56]: >AUTO_TECHSUPPORT_GLOBAL [](http://example.com/codeflow?start=4&length=23)

Do you have a strong reason to use UPPER_CASE function name? We prefer lower_case.

Comment (vivekreddynv) [reply show/plugins/auto_techsupport.py:56]: That's auto_generated by SONiC Auto CLI gen. https://github.com/Azure/sonic-utilities/pull/1644

File: scripts/coredump-compress (hunk 6 above)
Comment (qiluo-msft) [reply scripts/coredump-compress:26]: Then check python3* only? you are checking non-2.

Comment (vivekreddynv) [reply scripts/coredump-compress]: handled

**Comment by dgsudharsan** [comment]
@qiluo-msft  Can you please sign off if all comments are addressed?

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**